c/compress_similar_files/compress_similar_files.c

```diff
@@ -0,0 +1,291 @@
+          if (ref.position - ref.distance + ref.copy_len - 1 >= end + 5) {
+            new_position = end + ref.distance - (end - start);
+            new_distance = ref.distance;
+            new_copy_len = ref.copy_len - (end - ref.position + ref.distance);
+            new_max_distance = MIN(ref.max_distance, new_position);
+            *(*new_backward_references + new_back_ref_size) = (BackwardReferenceFromDecoder){new_position, new_copy_len, new_distance, new_max_distance};
+            new_back_ref_size++;
+          }
+        }
+        /* If the equal substring starts before removing content
+           then look where it ends */
+        else if (ref.position - ref.distance < start) {
+          if (ref.position - ref.distance + ref.copy_len - 1 < start) {
+            new_distance = ref.distance - (end - start);
+            new_position = ref.position - (end - start);
+            new_max_distance = MIN(ref.max_distance, new_position);
+            *(*new_backward_references + new_back_ref_size) = (BackwardReferenceFromDecoder){new_position, ref.copy_len, new_distance, new_max_distance};
+            new_back_ref_size++;
+          }
+          /* If the end is after removing content then cut the end
+             and save if the new copy len is long enough */
+          else {
+            new_position = ref.position - (end - start);
+            new_distance = ref.distance - (end - start);
+            new_copy_len = start - (ref.position - ref.distance);
+            new_max_distance = MIN(ref.max_distance, new_position);
+            if (new_copy_len >= 3) {
+              *(*new_backward_references + new_back_ref_size) = (BackwardReferenceFromDecoder){new_position, new_copy_len, new_distance, new_max_distance};
+              new_back_ref_size++;
+            }
+          }
+        }
+        /* If the equal substring starts after removing content
+           then just save it */
+        else {
+          new_position = ref.position - (end - start);
+          new_max_distance = MIN(ref.max_distance, new_position);
+          *(*new_backward_references + new_back_ref_size) = (BackwardReferenceFromDecoder){new_position, ref.copy_len, ref.distance, new_max_distance};
+          new_back_ref_size++;
+        }
+      }
+    }
+  }
+  return new_back_ref_size;
+}
+
+/* Saves a block <pos_begin, pos_end, old_type> into new_block_splits*/
+void SaveNewBlock(size_t pos_begin, size_t pos_end, uint8_t old_type,
+                  int* types_mapping, BlockSplitFromDecoder* new_block_splits,
+                  size_t* new_num_blocks, size_t* new_num_types) {
+  /* Haven't seen this type before, save a mapping */
+  if (types_mapping[old_type] == -1) {
+    types_mapping[old_type] = *new_num_types;
+  }
+  /* Create a new block */
+  if (*new_num_blocks == 0 || (new_block_splits->types[*new_num_blocks - 1] != types_mapping[old_type] && pos_end - pos_begin > 3)) {
```

Comment:
They actually can't exceed the memory I allocated for them (even in case of bad input). I allocated block_splits->num_blocks and then did a saving in those arrays less or equal than block_splits->num_blocks times. But I guess, wouldn't hurt to check.

Reply:
In that case, a BROTLI_DCHECK would be a good way to make sure that future callers of this function keep those constraints in mind